substreams/ethereum-maverick/src/modules.rs

```diff
@@ -0,0 +1,456 @@
+        0,
+        &map.tx_components
+            .iter()
+            .flat_map(|tx_components| &tx_components.components)
+            .map(|component| format!("pool:{0}", component.id))
+            .collect::<Vec<_>>(),
+        1,
+    );
+}
+
+/// Simply stores the `ProtocolComponent`s with the pool id as the key
+#[substreams::handlers::store]
+pub fn store_component_tokens(map: BlockTransactionProtocolComponents, store: StoreSetString) {
+    map.tx_components
+        .iter()
+        .flat_map(|tx_components| &tx_components.components)
+        .for_each(|component| {
+            store.set(
+                0,
+                format!("pool:{0}", component.id),
+                &component
+                    .tokens
+                    .iter()
+                    .map(|token| hex::encode(token))
+                    .join(":".into()),
+            );
+        });
+}
+
+/// Since the `Swap`, `AddLiquidity`, `RemoveLiuidity` events administer only deltas, we need to
+/// leverage a map and a store to be able to tally up final balances for tokens in a pool.
+#[substreams::handlers::map]
+pub fn map_relative_balances(
+    block: eth::v2::Block,
+    pools_store: StoreGetInt64,
+    tokens_store: StoreGetString,
+) -> Result<BlockBalanceDeltas, anyhow::Error> {
+    let deltas = block
+        .logs()
+        .filter(|log| {
+            pools_store
+                .get_last(format!("pool:{0}", hex::encode(&log.address())))
+                .is_some()
+        })
+        .flat_map(|log| {
+            if let Some(event) = abi::pool::events::Swap::match_and_decode(log) {
+                let tokens = tokens_store
+                    .get_last(format!("pool:{}", hex::encode(log.address())))
+                    .unwrap()
+                    .split(":")
+                    .map(|token| token.to_owned()) // Clone the tokens
+                    .collect::<Vec<_>>();
+                let (token_a, token_b) = if event.token_a_in {
+                    (
+                        hex::decode(tokens[0].clone()).unwrap(),
+                        hex::decode(tokens[1].clone()).unwrap(),
+                    )
+                } else {
+                    (
+                        hex::decode(tokens[1].clone()).unwrap(),
+                        hex::decode(tokens[0].clone()).unwrap(),
+                    )
+                };
+                vec![
+                    BalanceDelta {
+                        ord: log.log.ordinal,
+                        tx: Some(tx_from_log(&log)),
+                        token: token_a,
+                        delta: event.amount_in.to_signed_bytes_be(),
+                        component_id: log.address().into(),
+                    },
+                    BalanceDelta {
+                        ord: log.log.ordinal,
+                        tx: Some(tx_from_log(&log)),
+                        token: token_b,
+                        delta: event.amount_out.to_signed_bytes_be(),
```

Comment:
Please verify if this produces the correct sign (I assume it should be negative)